src/main/java/org/embulk/util/rubytime/ParserWithContext.java

```diff
@@ -217,71 +218,57 @@ private void consumeMonthOfYearName(final Parsed.Builder builder) {
+            builder.setDayOfMonth(this.consumeDigitsInInt(2, 1, 31, "invalid day of month"));
         }
-
-        this.throwIfOutOfRange(dayOfMonth, 1, 31, "invalid day of month");
-        builder.setDayOfMonth((int) dayOfMonth);
     }
 
     private void consumeWeekBasedYearWithCentury(final Parsed.Builder builder, final FormatToken nextToken) {
         if (isNumberPattern(nextToken)) {
-            builder.setWeekBasedYear((int) this.consumeDigits(4));
+            builder.setWeekBasedYear(this.consumeDigitsInInt(4, 0, 9999, "invalid year"));
         } else {
-            builder.setWeekBasedYear((int) this.consumeDigitsMax());
+            // JSR-310 accepts only [-999_999_999, 999_999_999] as a year, not [Integer.MIN_VALUE, Integer.MAX_VALUE].
+            // It is different from Ruby's Date._strptime / Time.strptime.
+            builder.setWeekBasedYear(this.consumeDigitsInInt(Integer.MAX_VALUE, 0, Year.MAX_VALUE, "invalid year"));
         }
     }
 
     private void consumeWeekBasedYearWithoutCentury(final Parsed.Builder builder) {
-        final long weekBasedYearWithoutCentury = this.consumeDigits(2);
-
-        this.throwIfOutOfRange(weekBasedYearWithoutCentury, 0, 99, "invalid year");
-        builder.setWeekBasedYearWithoutCentury((int) weekBasedYearWithoutCentury);
+        builder.setWeekBasedYearWithoutCentury(this.consumeDigitsInInt(2, 0, 99, "invalid year"));
     }
 
     private void consumeHourOfDay(final Parsed.Builder builder) {
-        final long hourOfDay;
         if (isBlank(this.text, this.pos)) {
             this.pos += 1;  // Blank
-            hourOfDay = this.consumeDigits(1);
+            builder.setHour(this.consumeDigitsInInt(1, 0, 24, "invalid hour of day"));
         } else {
-            hourOfDay = this.consumeDigits(2);
+            builder.setHour(this.consumeDigitsInInt(2, 0, 24, "invalid hour of day"));
         }
-
-        this.throwIfOutOfRange(hourOfDay, 0, 24, "invalid hour of day");
-        builder.setHour((int) hourOfDay);
     }
 
     private void consumeHourOfAmPm(final Parsed.Builder builder) {
-        final long hourOfAmPm;
         if (isBlank(this.text, this.pos)) {
             this.pos += 1;  // Blank
-            hourOfAmPm = this.consumeDigits(1);
+            builder.setHour(this.consumeDigitsInInt(1, 1, 12, "invalid hour of am/pm"));
         } else {
-            hourOfAmPm = this.consumeDigits(2);
+            builder.setHour(this.consumeDigitsInInt(2, 1, 12, "invalid hour of am/pm"));
         }
-
-        this.throwIfOutOfRange(hourOfAmPm, 1, 12, "invalid hour of am/pm");
-        builder.setHour((int) hourOfAmPm);
     }
 
     private void consumeDayOfYear(final Parsed.Builder builder) {
-        final long dayOfYear = this.consumeDigits(3);
-
-        this.throwIfOutOfRange(dayOfYear, 1, 365, "invalid day of year");
-        builder.setDayOfYear((int) dayOfYear);
+        builder.setDayOfYear(this.consumeDigitsInInt(3, 1, 365, "invalid day of year"));
```

Comment:
Don't accept leap year that has 366 days?

Reply:
Ahh, really a good catch. Thanks!

Original Ruby has done it with `366`.
https://svn.ruby-lang.org/cgi-bin/viewvc.cgi/tags/v2_5_1/ext/date/date_strptime.c?annotate=63029#l341

But Embulk's Ruby strptime parser had been doing it with `365` since the first implementation in v0.8.27.
https://github.com/embulk/embulk/blob/v0.8.27/embulk-jruby-strptime/src/main/java/org/embulk/spi/time/StrptimeParser.java#L519

We should fix it. This rubytime library would be fixed soon. The Embulk core will be fixed in later versions by applying this library.

Reply:
@sakama Fixed it in 45755883dee0615dca82164e2bf10d34af8914ff.

Reply:
I also didn't notice it at previous PR 👍